File: 2021/11/03/domino_pairs.py
"""
https://leetcode.com/problems/number-of-equivalent-domino-pairs/
Author: Andrew Jarombek
Date: 11/3/2021
"""

from typing import List


def num_equiv_domino_pairs(dominoes: List[List[int]]) -> int:
    pairs = 0

    domino_map = {}

    for domino in dominoes:
        sorted_domino = (min(domino[0], domino[1]), max(domino[0], domino[1]))
        count = domino_map.get(sorted_domino, 0)
        pairs += count
        domino_map[sorted_domino] = count + 1

    return pairs

File: 2022/10/01/dominoPairs.ts
/**
 * https://leetcode.com/problems/number-of-equivalent-domino-pairs/
 * @author Andrew Jarombek
 * @since 10/1/2022
 */

function numEquivDominoPairs(dominoes: number[][]): number {
    let map = new Map<string, number>();
    let result = 0;

    for (let domino of dominoes) {
        let [x, y] = domino;
        let id = `${Math.min(x, y)},${Math.max(x, y)}`;
        let val = map.get(id) ?? 0;
        result += val;
        map.set(id, val + 1);
    }

    return result;
}